### PR TITLE
Optimizations for listing files in directory

### PR DESCRIPTION
**Pull request by jdpurcell**
This addresses slowness when switching images as discussed in #476, #509, #514. While there may be more that could be done, and I only tested with a moderately sized folder (~1500 images), I think this is a good first step that noticeably speeds things up with some relatively simple and (IMO) safe code changes.

### CODE REVIEW

File: src/qvimagecore.cpp
Comment (jdpurcell) [inline src/qvimagecore.cpp:162]: When switching between images in a folder, `goToFile` has already called `updateFolderInfo` shortly before `loadPixmap` runs, so this ends up being mostly redundant aside from updating the loaded image index. We only need to run a full update if the image wasn't found in the cached list, as may happen when opening an image from a different folder (e.g. with File > Open).

File: src/qvimagecore.cpp
Comment (jdpurcell) [inline src/qvimagecore.cpp:239]: I'm not sure why but using `QDirIterator` directly (instead of `entryInfoList()` which relies on `QDirIterator` behind the scenes) resulted in a significant performance increase when testing on Windows. On Mac, the `entryInfoList()` call was already fast to begin with, and this change didn't affect performance. I passed in Filter::Files because the existing code could add folders to `fileInfoList`, for example if a folder was named like "test.jpg".

File: src/qvimagecore.cpp
Comment (jdpurcell) [inline src/qvimagecore.cpp:268]: Calling `fileInfo.dir().count()` results in `QDir` populating the entries behind the scenes. Instead, I'm just calling `getCompatibleFiles()` first and using its count to avoid the redundant file system access.

File: src/qvimagecore.cpp
Comment (jdpurcell) [inline src/qvimagecore.cpp:237]: Passing `QDir::SortFlag::Unsorted` forces it to skip the sort step, versus `QDir::SortFlag::NoSort` which ends up sorting it based on some default rules. Note that if you saw my previous commit where I accomplished something similar using `QDirIterator`, I ended up reverting that due to inconsistent performance when testing on a larger set of files.

**Comment by jdpurcell** [comment]
My latest commit takes care of the remaining slowness that occurred when sorting by Last Modified or Type. Thanks to [AlttiRi](https://github.com/AlttiRi) for [pointing this out](https://github.com/jurplel/qView/issues/476#issuecomment-1024971963).

I think this is about the limit of what can be done with the current approach. I find it perfectly usable. If one wanted to optimize things further, QFileSystemWatcher triggering background updates would be worth trying. I may try it myself even, but that's outside the scope of what I wanted to accomplish with this pull request.

File: src/settingsmanager.cpp
Comment (jdpurcell) [inline src/settingsmanager.cpp:179]: Defaults to false to preserve existing behavior. When Mac is accessing a Windows network share, I believe it's doing a round trip for every file to check isHidden(), so toggling this setting this helps massively.

File: src/settingsmanager.cpp
Comment (jdpurcell) [inline src/settingsmanager.cpp:180]: Defaults to true to preserve existing behavior. If a file extension isn't recognized, or sometimes even if it is but could refer to multiple mime types, QMimeDatabase will inspect the file contents. This of course creates additional I/O, especially noticeable with large network shares, and toggling this setting off can help massively.

File: src/qvimagecore.cpp
Comment (jdpurcell) [inline src/qvimagecore.cpp:262]: I intentionally passed in `absoluteFilePath` instead of `fileInfo` here to work around a performance issue wherein QMimeDatabase will check isDir() if a QFileInfo is passed in (so it can return "inode/directory" when appropriate), even if the MatchExtension mode was requested.

**Comment by jdpurcell** [comment]
I added some settings that can greatly help performance with large network shares. I'll leave it up to jurplel to decide if and where they should go in the UI.

**Comment by jurplel** [comment]
I am aware of this PR, will review soon, thank you for your work :)

**Comment by jdpurcell** [comment]
As for the new settings, I would propose that:
1) MIME content detection is off by default since it was added only recently (after 4.0 as part of #396) and has a large potential to impact performance.
2) The UI looks like:
image

This is accomplished in my latest commit. If you're fine with that, unless you find anything during review, it's all good to go! If you have other opinions for the settings, feel free to ask me to make changes or make edits yourself as appropriate.

**Review by jurplel — CHANGES_REQUESTED** [review]

**Comment by jdpurcell** [comment]
@jurplel I just tested and the performance hit is identical whether omitting `QDir::Hidden`, or checking `isHidden` in the loop afterwards. To be clear, I've only found this to matter when viewing a Windows file share from macOS. Another Windows client viewing the same share has no problem, maybe it's using a newer version of SMB or the macOS implementation has some limitations. But this folder, with ~16K files, takes ~100ms to list all files on macOS, vs a full two minutes if asked to filter out the hidden files.

You could have it potentially do a delayed check on isHidden() right before navigating to a file, but that would of course throw off the file count in the titlebar and have the index jump by more than 1 if it ends up skipping over a hidden file.

**Comment by jdpurcell** [comment]
Or it could always show hidden images? Would anyone really have a folder with mixed hidden and non-hidden images? 😄

**Comment by jurplel** [comment]
> Or it could always show hidden images? Would anyone really have a folder with mixed hidden and non-hidden images? smile

Let's do that for now

**Comment by jdpurcell** [comment]
@jurplel Okay, so I should remove that setting (as opposed to keeping it but it defaulting to on)?

**Comment by jurplel** [comment]
> @jurplel Okay, so I should remove that setting (as opposed to keeping it but it defaulting to on)?
> 
> EDIT: Proceeding this way, let me know if you meant otherwise.

Yeah I think that sounds good for now until someone requests otherwise or until i redo the settings page and find a better way to educate users on the consequences of this feature

**Review by jurplel — APPROVED** [review]

**Pull request merged**
